File: src/main/java/uk/gov/justice/hmpps/offenderevents/services/ReleasePrisonerReasonCalculator.java
package uk.gov.justice.hmpps.offenderevents.services;

import org.springframework.stereotype.Component;

@Component
public class ReleasePrisonerReasonCalculator {
    private final PrisonApiService prisonApiService;

    public ReleasePrisonerReasonCalculator(PrisonApiService prisonApiService) {
        this.prisonApiService = prisonApiService;
    }


    public ReleaseReason calculateReasonForRelease(String offenderNumber) {
        final var prisonerDetails = prisonApiService.getPrisonerDetails(offenderNumber);
        final var currentLocation = prisonerDetails.currentLocation();
        final var currentPrisonStatus = prisonerDetails.currentPrisonStatus();
        final var latestLocationId = prisonerDetails.latestLocationId();
        final var nomisMovementReason = new MovementReason(prisonerDetails.lastMovementReasonCode());

        final var reasonWithDetails =
            switch (prisonerDetails.typeOfMovement()) {
                case TEMPORARY_ABSENCE -> new ReasonWithDetails(Reason.TEMPORARY_ABSENCE_RELEASE, new MovementReason(prisonerDetails.lastMovementReasonCode()));
                case COURT -> new ReasonWithDetails(Reason.SENT_TO_COURT, new MovementReason(prisonerDetails.lastMovementReasonCode()));
                case TRANSFER -> new ReasonWithDetails(Reason.TRANSFERRED, new MovementReason(prisonerDetails.lastMovementReasonCode()));
                case RELEASED -> {
                    final var reason = switch (prisonerDetails.movementReason()) {
                        case HOSPITALISATION -> Reason.RELEASED_TO_HOSPITAL;
                        default -> Reason.RELEASED;
                    };
                    yield new ReasonWithDetails(reason, String.format("Movement reason code %s", prisonerDetails.lastMovementReasonCode()), new MovementReason(prisonerDetails.lastMovementReasonCode()));
                }
                default -> new ReasonWithDetails(Reason.UNKNOWN, String.format("Movement type code %s", prisonerDetails.lastMovementTypeCode()), new MovementReason(prisonerDetails.lastMovementReasonCode()));
            };

        return new ReleaseReason(reasonWithDetails.reason(), reasonWithDetails.details(), currentLocation, currentPrisonStatus, latestLocationId, nomisMovementReason);
    }

    enum Reason {
        UNKNOWN,
        TEMPORARY_ABSENCE_RELEASE,
        RELEASED_TO_HOSPITAL,
        RELEASED,
        SENT_TO_COURT,
        TRANSFERRED
    }

    record MovementReason(String code) {
    }

    record ReasonWithDetails(Reason reason, String details, MovementReason nomisMovementReason) {
        public ReasonWithDetails(Reason reason, MovementReason nomisMovementReason) {
            this(reason, null, nomisMovementReason);
        }
    }

    record ReleaseReason(Reason reason, String details, CurrentLocation currentLocation, CurrentPrisonStatus currentPrisonStatus, String prisonId, MovementReason nomisMovementReason) implements PrisonerMovementReason {
        public ReleaseReason(Reason reason, CurrentLocation currentLocation, CurrentPrisonStatus currentPrisonStatus, String prisonId, MovementReason nomisMovementReason) {
            this(reason, null, currentLocation, currentPrisonStatus, prisonId, nomisMovementReason);
        }

        public boolean hasPrisonerActuallyBeenRelease() {
            return currentLocation != CurrentLocation.IN_PRISON;
        }
    }
}

File: src/main/java/uk/gov/justice/hmpps/offenderevents/services/ReleasePrisonerReasonCalculator.kt
package uk.gov.justice.hmpps.offenderevents.services

import org.springframework.stereotype.Component
import uk.gov.justice.hmpps.offenderevents.services.MovementReason.HOSPITALISATION

@Component
class ReleasePrisonerReasonCalculator(private val prisonApiService: PrisonApiService) {
  internal fun calculateReasonForRelease(offenderNumber: String): ReleaseReason {
    val prisonerDetails = prisonApiService.getPrisonerDetails(offenderNumber)
    val currentLocation = prisonerDetails.currentLocation()
    val currentPrisonStatus = prisonerDetails.currentPrisonStatus()
    val latestLocationId = prisonerDetails.latestLocationId
    val nomisMovementReason = MovementReason(prisonerDetails.lastMovementReasonCode)
    val reasonWithDetails = when (prisonerDetails.typeOfMovement()) {
      MovementType.TEMPORARY_ABSENCE -> ReasonWithDetails(
        reason = Reason.TEMPORARY_ABSENCE_RELEASE,
        nomisMovementReason = MovementReason(prisonerDetails.lastMovementReasonCode),
      )

      MovementType.COURT -> ReasonWithDetails(
        reason = Reason.SENT_TO_COURT,
        nomisMovementReason = MovementReason(prisonerDetails.lastMovementReasonCode),
      )

      MovementType.TRANSFER -> ReasonWithDetails(
        Reason.TRANSFERRED,
        nomisMovementReason = MovementReason(prisonerDetails.lastMovementReasonCode),
      )

      MovementType.RELEASED -> {
        val reason = when (prisonerDetails.movementReason()) {
          HOSPITALISATION -> Reason.RELEASED_TO_HOSPITAL
          else -> Reason.RELEASED
        }
        ReasonWithDetails(
          reason,
          "Movement reason code ${prisonerDetails.lastMovementReasonCode}",
          MovementReason(prisonerDetails.lastMovementReasonCode),
        )
      }

      else -> ReasonWithDetails(
        Reason.UNKNOWN,
        "Movement type code ${prisonerDetails.lastMovementTypeCode}",
        MovementReason(prisonerDetails.lastMovementReasonCode),
      )
    }
    return ReleaseReason(
      reasonWithDetails.reason,
      reasonWithDetails.details,
      currentLocation,
      currentPrisonStatus,
      latestLocationId,
      nomisMovementReason,
    )
  }

  internal enum class Reason {
    UNKNOWN, TEMPORARY_ABSENCE_RELEASE, RELEASED_TO_HOSPITAL, RELEASED, SENT_TO_COURT, TRANSFERRED
  }

  internal data class MovementReason(val code: String)

  internal data class ReasonWithDetails(
    val reason: Reason,
    val details: String? = null,
    val nomisMovementReason: MovementReason,
  )

  internal data class ReleaseReason(
    val reason: Reason,
    override val details: String? = null,
    override val currentLocation: CurrentLocation,
    override val currentPrisonStatus: CurrentPrisonStatus,
    override val prisonId: String,
    val nomisMovementReason: MovementReason,
  ) : PrisonerMovementReason {
    fun hasPrisonerActuallyBeenRelease(): Boolean = currentLocation != CurrentLocation.IN_PRISON
  }
}
